compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -2843,6 +2857,92 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// Try to match an implementation of `Index` against a self type, and report
+    /// the unsatisfied predicates that result from confirming this impl.
+    ///
+    /// Given an index expression, sometimes the `Self` type shallowly but does not
+    /// deeply satisfy an impl predicate. Instead of simply saying that the type
+    /// does not support being indexed, we want to point out exactly what nested
+    /// predicates cause this to be, so that the user can add them to fix their code.
+    fn find_and_report_unsatisfied_index_impl(
+        &self,
+        index_expr_hir_id: HirId,
+        base_expr: &hir::Expr<'_>,
+        base_ty: Ty<'tcx>,
+    ) -> Option<(ErrorGuaranteed, Ty<'tcx>, Ty<'tcx>)> {
+        let Some(index_trait_def_id) = self.tcx.lang_items().index_trait() else {
+            return None;
+        };
+
+        let mut relevant_impls = vec![];
+        self.tcx.for_each_relevant_impl(index_trait_def_id, base_ty, |impl_def_id| {
+            relevant_impls.push(impl_def_id);
+        });
+
+        let impl_def_id = match &relevant_impls[..] {
+            // Only report unsatisfied impl predicates if there's one impl
+            [impl_def_id] => *impl_def_id,
+            _ => return None,
+        };
```

Comment:
Would a let-else work?
```suggestion
        // Only report unsatisfied impl predicates if there's one impl.
        let &[impl_def_id] = match &relevant_impls[..] else { return None };
```